### PR TITLE
Use Redis charm library to access Redis connection info

### PR DESCRIPTION
**Pull request by weiiwang01**
Switch to use the Redis charm library (`charms.redis_k8s.v0.redis`) to access Redis connection info from Redis integration data.

### CODE REVIEW

File: xiilib/databases.py [hunk 1]
```diff
@@ -54,16 +56,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str):
             charm: The requesting charm object.
             relation_name: The redis relation name.
         """
-        self._charm = charm
-        self.relation_name = relation_name
-        # redis charm doesn't provide database id via the relation
+        # we don't need store for Redis
```
Comment (jdkandersson) [inline xiilib/databases.py]: What is the store used for please?

File: xiilib/_gunicorn/charm.py [hunk 2]
```diff
@@ -34,6 +39,8 @@ def get_wsgi_config(self) -> BaseModel:
     def get_cos_dir(self) -> str:
         """Return the directory with COS related files."""
 
+    on = RedisRelationCharmEvents()
```
Comment (jdkandersson) [inline xiilib/_gunicorn/charm.py:43]: Why is this required here please? This only seems to be there for the redis integration and not other integration?

File: xiilib/databases.py [hunk 3]
```diff
@@ -54,16 +56,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str):
...
+        self._redis = RedisRequires(charm=charm, _stored=MagicMock(), relation_name=relation_name)
         self.database = ""
-        self.on = self._RedisDatabaseRequiresEventShim(
-            self._charm.on[relation_name].relation_changed
-        )
+        self.relation_name = relation_name
+        relation = charm.model.get_relation(relation_name)
+        self.relation_id = relation.id if relation else None
+        self.on = self._RedisDatabaseRequiresEventShim(charm.on.redis_relation_updated)
```
Comment (jdkandersson) [inline xiilib/databases.py]: I know this will be a point of discussion with @gregory-schiano, could you please describe the benefits of using the shim here compared to implementing the relation without using the databases module? This will help kick start the discussion with Gregory

File: xiilib/databases.py [hunk 4]
```diff
@@ -54,16 +56,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str):
...
 
-    def fetch_relation_data(
-        self, fields: typing.List[str]
+    def fetch_relation_data(  # argument fields is used for matching the function signature
```
Comment (jdkandersson) [inline xiilib/databases.py]: What function signature is being matched here please?

File: xiilib/databases.py [hunk 5]
```diff
@@ -73,20 +75,9 @@ def fetch_relation_data(
...
-            for unit in relation.units:
-                if not (hostname := relation.data[unit].get("hostname")):
-                    continue
-                if not (port := relation.data[unit].get("port")):
-                    continue
-                endpoints.append(f"{hostname}:{port}")
-            if endpoints:
-                data[relation.id] = (
-                    {"endpoints": ",".join(endpoints)} if "endpoints" in fields else {}
-                )
```
Comment (jdkandersson) [inline xiilib/databases.py:89]: Is this no longer required for all the databases we support?

File: xiilib/databases.py (hunk 1 above)
Comment (weiiwang01) [reply xiilib/databases.py]: This is required by the Redis charm library, it doesn't have any use for our use case.

File: xiilib/_gunicorn/charm.py (hunk 2 above)
Comment (weiiwang01) [reply xiilib/_gunicorn/charm.py:43]: This is required by the Redis charm library.

File: xiilib/databases.py (hunk 4 above)
Comment (weiiwang01) [reply xiilib/databases.py]: It's for matching `charms.data_platform_libs.v0.data_interfaces.DataRelation.fetch_relation_data`.

File: xiilib/databases.py (hunk 5 above)
Comment (weiiwang01) [reply xiilib/databases.py:89]: This is only used for redis so it can be removed now.

File: xiilib/databases.py (hunk 3 above)
Comment (weiiwang01) [reply xiilib/databases.py]: If we keep the Redis handling within the database module, it can share all the database handling logic inside the same module. For example, check if the database is declared in the metadata, block operations if a non-optional database is not ready, generate the database environment variables, and use the same database event naming convention.

File: xiilib/_gunicorn/charm.py (hunk 2 above)
Comment (jdkandersson) [reply xiilib/_gunicorn/charm.py:43]: What does the redis charm library use it for please?

Comment (weiiwang01) [reply xiilib/_gunicorn/charm.py:43]: The Redis charm library uses it for Redis related events.

File: xiilib/databases.py (hunk 4 above)
Comment (jdkandersson) [reply xiilib/databases.py]: What is the reason for needing to match it please?

File: xiilib/_gunicorn/charm.py (hunk 2 above)
Comment (weiiwang01) [reply xiilib/_gunicorn/charm.py:43]: It's just how the Redis charm library is designed.

File: xiilib/databases.py (hunk 1 above)
Comment (jdkandersson) [reply xiilib/databases.py]: So can we pass it the real storage even if we don't use it? Generally production code shouldn't have a mock in it since that is intended for use during testing

File: xiilib/databases.py (hunk 5 above)
Comment (jdkandersson) [reply xiilib/databases.py:89]: The function is called `fetch_relation_data` which seems to indicate that it fetches relation data for all the databases?

File: xiilib/databases.py (hunk 1 above)
Comment (weiiwang01) [reply xiilib/databases.py]: `StoreState` is deprecated I think so I would prefer not using that

File: xiilib/databases.py (hunk 4 above)
Comment (weiiwang01) [reply xiilib/databases.py]: Well, the shim needs to match the function it replaced; that's the definition of a shim.

File: xiilib/databases.py (hunk 1 above)
Comment (jdkandersson) [reply xiilib/databases.py]: We are forced to use it by the library, even if it is deprecated it is better than passing a mock

File: xiilib/_gunicorn/charm.py (hunk 2 above)
Comment (jdkandersson) [reply xiilib/_gunicorn/charm.py:43]: Ok, should this be a private attribute then?

Comment (weiiwang01) [reply xiilib/_gunicorn/charm.py:43]: No it has to be called `on`

File: xiilib/databases.py (hunk 1 above)
Comment (weiiwang01) [reply xiilib/databases.py]: okay, if that makes you happy

File: xiilib/databases.py [hunk 6]
```diff
@@ -49,20 +48,23 @@ def __init__(self, relation_changed: typing.Callable):
...
-        self._redis = RedisRequires(charm=charm, _stored=MagicMock(), relation_name=relation_name)
+        super().__init__(charm, "redis-shim")
+        self._store.set_default(redis_relation={})
+        self._redis = RedisRequires(charm=charm, _stored=self._store, relation_name=relation_name)
         self.database = ""
         self.relation_name = relation_name
         relation = charm.model.get_relation(relation_name)
         self.relation_id = relation.id if relation else None
-        self.on = self._RedisDatabaseRequiresEventShim(charm.on.redis_relation_updated)
+        setattr(self, "on", self._RedisDatabaseRequiresEventShim(charm.on.redis_relation_updated))
```
Comment (jdkandersson) [inline xiilib/databases.py]: What is the reason for using `setattr` here rather than the assignment please?

Comment (weiiwang01) [reply xiilib/databases.py]: Problems caused by linter

Comment (jdkandersson) [reply xiilib/databases.py]: What is the problem please?

Comment (weiiwang01) [reply xiilib/databases.py]: Write to read-only attribute.

Comment (jdkandersson) [reply xiilib/databases.py]: And do you know why the linter believes it is a read only attribute?

File: xiilib/_gunicorn/charm.py [hunk 7]
```diff
@@ -59,12 +61,21 @@ def __init__(self, framework: ops.Framework, wsgi_framework: str) -> None:
             self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
             return
 
+        requires = load_requires()
+        if "redis" in requires and requires["redis"]["interface"] == "redis":
```
Comment (jdkandersson) [inline xiilib/_gunicorn/charm.py:65]: Should this be wrapped into a `try ... except` in case of a malformed `charmcraft.yaml`? Not sure if that would be caught by charmcraft before the build step

**Review by jdkandersson — COMMENTED** [review]
Looks good, thank you for updating this

File: xiilib/_gunicorn/charm.py (hunk 7 above)
Comment (jdkandersson) [reply xiilib/_gunicorn/charm.py:65]: specifically, for example, if `redis` is defined but perhaps doesn't have the `interface` key?

**Comment by weiiwang01** [comment]
Close this PR as content has been migrated to the new repo.
